chalk-engine/src/logic.rs

```diff
@@ -238,23 +236,35 @@ impl<C: Context> Forest<C> {
     /// In terms of the NFTD paper, this corresponds to the *Program
     /// Clause Resolution* step being applied eagerly, as many times
     /// as possible.
-    fn push_initial_strands(&mut self, context: &impl ContextOps<C>, table: TableIndex) {
+    fn build_table(
+        context: &impl ContextOps<C>,
+        table_idx: TableIndex,
+        goal: C::UCanonicalGoalInEnvironment,
+    ) -> Table<C> {
         // Instantiate the table goal with fresh inference variables.
-        let table_goal = self.tables[table].table_goal.clone();
+        let table = Table::<C>::new(goal.clone(), context.is_coinductive(&goal));
```

Comment:
You can just move this to `push_intial_strands_instantiated` (or better, just move that function *into* this one. We don't really need the distinction when this function is 4 lines long.